src/CodeGen_ARM.cpp

```diff
@@ -1065,6 +1062,165 @@ void CodeGen_ARM::visit(const LE *op) {
     CodeGen_Posix::visit(op);
 }
 
+void CodeGen_ARM::codegen_vector_reduce(const VectorReduce *op, const Expr &init) {
+    if (neon_intrinsics_disabled() ||
+        op->op == VectorReduce::Or ||
+        op->op == VectorReduce::And ||
+        op->op == VectorReduce::Mul) {
+        CodeGen_Posix::codegen_vector_reduce(op, init);
+        return;
+    }
+
+    // ARM has a variety of pairwise reduction ops for +, min,
+    // max. The versions that do not widen take two 64-bit args and
+    // return one 64-bit vector of the same type. The versions that
+    // widen take one arg and return something with half the vector
+    // lanes and double the bit-width.
+
+    int factor = op->value.type().lanes() / op->type.lanes();
+    if ((op->type.is_int() ||
```

Comment:
style nit: this expr is really hard to read. Might be better to pull out as a separate expr, perhaps even like

bool has_reduce = is_int || is_uint || is_float;
if (!case1) has_reduce = false;  // reason why
if (!case2) has_reduce = false;  // reason why

etc